File: BOJ/[1509] 팰린드롬 분할/woong-jae/1509.js
const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const input = [];
rl.on('line', function (line) {
    input.push(line.split(' '));
}).on("close", function () {
    console.log(solution(input[0][0]));
    process.exit();
});

function solution(input) {
    const N = input.length;

    const isPalindrome = Array.from(Array(N), () => Array(N).fill(false));
    for(let mid = 0; mid < N; mid++) {
        let left = mid, right = mid;
        while(left >= 0 && right < N && input[left] === input[right]) {
            isPalindrome[left--][right++] = true;
        }
        
        left = mid, right = mid + 1;
        while(left >= 0 && right < N && input[left] === input[right]) {
            isPalindrome[left--][right++] = true;
        }
    }

    const dp = Array(N + 1).fill(Infinity);
    dp[N - 1] = 1;
    dp[N] = 0;
    for(let start = N - 2; start >= 0; start--) {
        for(let end = start; end < N; end++) {
            if(!isPalindrome[start][end]) continue;
            dp[start] = Math.min(dp[start], dp[end + 1] + 1);
        }
    }

    return dp[0];
}

File: BOJ/[1509] 팰린드롬 분할/Johoseong/1509.py
import sys
input = sys.stdin.readline

arr = input().rstrip()
N = len(arr)
dp = [[0] * N for _ in range(N)]
result = [N] * (N + 1)
result[N - 1] = 1
result[N] = 0

for i in range(N): # 길이 1이면 무조건 팰린드롬임
    dp[i][i] = 1
for l in range(1, N):
    for i in range(N - l - 1, -1, -1):
        if arr[i] == arr[i + l]:
            if l == 1: # 길이 2인 팰린드롬
                dp[i][i + 1] = 1
                continue
            if dp[i + 1][i + l - 1] == 1: # 길이 3 이상인 팰린드롬
                dp[i][i + l] = 1

for i in range(N - 2, -1, -1):
    for j in range(i, N):
        if dp[i][j] == 1:
            result[i] = min(result[i], result[j + 1] + 1)
        else:
            result[i] = min(result[i], result[i + 1] + 1)

print(result[0])